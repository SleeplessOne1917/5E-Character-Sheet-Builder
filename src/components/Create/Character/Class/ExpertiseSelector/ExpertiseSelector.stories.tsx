import { ComponentMeta, ComponentStory } from '@storybook/react';
import { Provider } from 'react-redux';
import { getTestStore } from '../../../../../redux/store';
import ExpertiseSelector from './ExpertiseSelector';

export default {
	title: 'Components/Create/Character/Class/ExpertiseSelector',
	component: ExpertiseSelector,
	decorators: [
		Story => (
			<Provider store={getTestStore()}>
				<Story />
			</Provider>
		)
	]
} as ComponentMeta<typeof ExpertiseSelector>;

const Template: ComponentStory<typeof ExpertiseSelector> = () => (
	<ExpertiseSelector />
);

export const Default = Template.bind({});
