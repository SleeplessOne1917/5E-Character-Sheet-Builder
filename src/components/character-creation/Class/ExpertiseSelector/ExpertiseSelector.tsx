import { useCallback } from 'react';
import { useAppDispatch, useAppSelector } from '../../../../hooks/reduxHooks';
import { setExpertiseProficiency } from '../../../../redux/features/classInfo';
import { SrdProficiencyItem } from '../../../../types/srd';
import Select from '../../../Select/Select/Select';
import classes from './ExpertiseSelector.module.css';

const ExpertiseSelector = () => {
	const proficiencies = useAppSelector(
		state => state.editingCharacter.proficiencies
	).filter(prof => prof.type === 'SKILLS' || prof.index === 'thieves-tools');

	const selectedExpertiseProficiencies =
		useAppSelector(
			state => state.editingCharacter.classInfo.expertiseProficiencies
		) ?? [];

	const dispatch = useAppDispatch();

	const handleChange = useCallback(
		(index: number, value: SrdProficiencyItem | null) => {
			dispatch(setExpertiseProficiency({ index, proficiency: value }));
		},
		[dispatch]
	);

	return (
		<div data-testid="expertise-selector" className={classes.container}>
			<h2 className={classes.heading}>Expertise</h2>
			<div className={classes['selected-text']}>
				{selectedExpertiseProficiencies.filter(prof => prof).length}/
				{selectedExpertiseProficiencies.length} Proficiencies Selected
			</div>
			<div className={classes['proficiencies-container']}>
				{selectedExpertiseProficiencies.map((prof, index) => (
					<div
						key={index}
						className={`${classes['proficiency-container']}${
							prof ? ` ${classes.selected}` : ''
						}`}
					>
						<Select
							fontSize="1.3rem"
							id="select-proficiency"
							label="Select Proficiency"
							options={[{ value: 'blank', label: '\u2014' }].concat(
								proficiencies
									.filter(
										prof =>
											!selectedExpertiseProficiencies.some(
												s => s?.index === prof.index
											) ||
											prof.index ===
												selectedExpertiseProficiencies[index]?.index
									)
									.map(prof => ({
										value: prof.index,
										label: prof.name.replace(/.*:\s*/, '')
									}))
							)}
							value={prof ? prof.index : 'blank'}
							onChange={value =>
								handleChange(
									index,
									value !== 'blank'
										? (proficiencies.find(
												p => p.index === (value as string)
										  ) as SrdProficiencyItem)
										: null
								)
							}
						/>
					</div>
				))}
			</div>
		</div>
	);
};

export default ExpertiseSelector;
