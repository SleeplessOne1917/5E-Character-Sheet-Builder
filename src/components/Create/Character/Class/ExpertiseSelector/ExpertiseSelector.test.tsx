import { render, screen } from '@testing-library/react';
import { composeStories } from '@storybook/react';
import * as stories from './ExpertiseSelector.stories';

const { Default } = composeStories(stories);

it('renders correctly', () => {
	render(<Default />);

	expect(screen.getByTestId('expertise-selector')).toMatchSnapshot();
});
