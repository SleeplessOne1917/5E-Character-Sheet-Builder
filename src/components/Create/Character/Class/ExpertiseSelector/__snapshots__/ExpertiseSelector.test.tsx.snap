// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<div
  class="container"
  data-testid="expertise-selector"
>
  <h2
    class="heading"
  >
    Expertise
  </h2>
  <div
    class="selected-text"
  >
    0
    /
    0
     Proficiencies Selected
  </div>
  <div
    class="proficiencies-container"
  />
</div>
`;
